### PR TITLE
Added connection retry logic to Driver

### LINKED ISSUE #14875: SQL Azure transient connection error when server unavailable
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 4.1.1
* Platform and Target: Windows, IIS (Azure App Service), SQL Server Azure, HTTPS

### What you did
I'm using SQL Server on Azure (not by my choice) and it works fine for our application, but we recently tried using an on-demand instance for the database to reduce costs. There's no issue with the application, except when the database is waking up from pausing. I currently set it up to pause after 1 hour of inactivity. When connecting after the database was 'paused', SQL Server throws an error saying to basically check again in a few moments.

I tried to get around this error by setting the `loginTimeout` option on the database connection to 600 (10 minutes), hoping CakePHP could wait until the server is available or retry the connection at some point.

### What happened
The following error is thrown by the SQL Server driver when the database is still waking up: 
`[PDOException] SQLSTATE[08001]: [Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: Timeout error [258]`

And the following can also be thrown: 
```Error: [Cake\Database\Exception\MissingConnectionException] Connection to Sqlserver could not be established: SQLSTATE[HY000]: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Database 'xxxxx' on server 'xxxxx.database.windows.net' is not currently available.  Please retry the connection later.  If the problem persists, contact customer support, and provide them the session tracing ID of '{xxxxx}'. in D:\home\site\wwwroot\vendor\cakephp\cakephp\src\Database\Driver.php on line 121

Caused by: [PDOException] SQLSTATE[HY000]: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Database 'xxxxx' on server 'xxxxx.database.windows.net' is not currently available.  Please retry the connection later.  If the problem persists, contact customer support, and provide them the session tracing ID of '{xxxxx}'. in D:\home\site\wwwroot\vendor\cakephp\cakephp\src\Database\Driver.php on line 118
```

### What you expected to happen
I noticed there is some connection retrying code in CakePHP, and I understand this is a pretty specific platform issue. I opened this issue to determine if this is something that either the framework can allow handling, if this is truly a bug (or unsupported), and if there is a workaround I can do. 

Thanks for all the great work on the framework!

### PR DESCRIPTION
**Pull request by othercorey**
Closes https://github.com/cakephp/cakephp/issues/14875

### CODE REVIEW

File: src/Database/Driver.php [hunk 1]
```diff
@@ -486,6 +519,16 @@ public function getMaxAliasLength(): ?int
         return static::MAX_ALIAS_LENGTH;
     }
 
+    /**
+     * Returns the number of connection retry attempts made.
+     *
+     * @return int
+     */
+    public function getLastConnectRetries(): int
```
Comment (ADmad) [inline src/Database/Driver.php:527]: This could be just `getConnectRetries()`.

Comment (othercorey) [reply src/Database/Driver.php]: Done. I also didn't update the retry count const name in SqlServer.

File: src/Database/Driver.php [hunk 2]
```diff
@@ -486,6 +519,16 @@ public function getMaxAliasLength(): ?int
         return static::MAX_ALIAS_LENGTH;
     }
 
+    /**
+     * Returns the number of connection retry attempts made.
+     *
+     * @return int
+     */
+    public function getConnectRetries(): int
```
Comment (ADmad) [inline src/Database/Driver.php:527]: Would also need to add `@method` annotation to `DriverInterface` for this new method.

**Comment by lorenzo** [comment]
We used to have a Retry class for this, which is used in Connectiin if I remember correctly. Would it be worth reusing it?

**Comment by othercorey** [comment]
> We used to have a Retry class for this, which is used in Connectiin if I remember correctly. Would it be worth reusing it?

CommandRetry wraps `Exception` while Driver wraps `PDOException`. Is that ok? We could maybe create a strategy in `connect` and pass it to `_connect`. Just depends how much we want to do here.

**Comment by othercorey** [comment]
I think `Connection::connect` double wraps `Driver::connect` to catch `Exception`.

**Comment by lorenzo** [comment]
I think a custom strategy may be a good idea

**Comment by ADmad** [comment]
Given the changes this should target next minor.

**Comment by othercorey** [comment]
This implements connect retry as a CommandRetry strategy.

Updated CommandRetry to pull the max retries config from the strategy.

File: src/Database/Retry/ConnectWaitStrategy.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    public function shouldRetry(Exception $exception): bool
+    {
+        if (
+            $exception instanceof PDOException &&
+            (
+                in_array($exception->getCode(), $this->errorCodes) ||
+                (
+                    $exception->errorInfo &&
+                    in_array($exception->errorInfo[1], $this->errorCodes)
```
Comment (othercorey) [inline src/Database/Retry/ConnectWaitStrategy.php:76]: This situation is caused by https://bugs.php.net/bug.php?id=64705. errorInfo is null when the exception is thrown from PDO construction.

File: src/Core/Retry/RetryStrategyInterface.php [hunk 4]
```diff
@@ -24,12 +24,18 @@
  */
 interface RetryStrategyInterface
 {
+    /**
+     * Returns the maximum number of retries allowed.
+     *
+     * @return int
+     */
+    public function getMaxRetries(): int;
```
Comment (markstory) [inline src/Core/Retry/RetryStrategyInterface.php:32]: Normally adding to interfaces is not backwards compatible. I doubt that this interface has been implemented by userland code though.

Comment (othercorey) [reply src/Core/Retry/RetryStrategyInterface.php:32]: `Cake\Database\Retry\ReconnectStrategy` is marked internal so I thought the whole mechanism was internal.

Comment (othercorey) [reply src/Core/Retry/RetryStrategyInterface.php:32]: Looking at the Orcale driver - the only other code I can think of to customize this - doesn't seem to have any strategy code.

**Comment by othercorey** [comment]
Re-implemented without breaking BC of CommandRetry and RetryStrategyInterface.

**Review by markstory — APPROVED** [review]
Looks good to me.

**Comment by othercorey** [comment]
Should this be for 4.1 patch for 4.2?

**Comment by markstory** [comment]
@othercorey I think this can fit in 4.1. It will be a no op for most people and no interfaces changed.

**Pull request merged**